modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/TruncateHitsResponseProcessor.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.search.pipeline.common;
+
+import org.opensearch.action.search.SearchRequest;
+import org.opensearch.action.search.SearchResponse;
+import org.opensearch.ingest.ConfigurationUtils;
+import org.opensearch.search.SearchHit;
+import org.opensearch.search.SearchHits;
+import org.opensearch.search.pipeline.AbstractProcessor;
+import org.opensearch.search.pipeline.Processor;
+import org.opensearch.search.pipeline.SearchResponseProcessor;
+import org.opensearch.search.pipeline.StatefulSearchResponseProcessor;
+import org.opensearch.search.pipeline.common.helpers.ContextUtils;
+import org.opensearch.search.pipeline.common.helpers.SearchResponseUtil;
+
+import java.util.Map;
+
+import static org.opensearch.search.pipeline.common.helpers.ContextUtils.applyContextPrefix;
+
+/**
+ * Truncates the returned search hits from the {@link SearchResponse}. If no target size is specified in the pipeline, then
+ * we try using the "original_size" value from the request context, which may have been set by {@link OversampleRequestProcessor}.
+ */
+public class TruncateHitsResponseProcessor extends AbstractProcessor implements StatefulSearchResponseProcessor {
+    /**
+     * Key to reference this processor type from a search pipeline.
+     */
+    public static final String TYPE = "truncate_hits";
+    static final String TARGET_SIZE = "target_size";
+    private final int targetSize;
+    private final String contextPrefix;
+
+    @Override
+    public String getType() {
+        return TYPE;
+    }
+
+    private TruncateHitsResponseProcessor(String tag, String description, boolean ignoreFailure, int targetSize, String contextPrefix) {
+        super(tag, description, ignoreFailure);
+        this.targetSize = targetSize;
+        this.contextPrefix = contextPrefix;
+    }
+
+    @Override
+    public SearchResponse processResponse(SearchRequest request, SearchResponse response, Map<String, Object> requestContext) {
+
+        int size;
+        if (targetSize < 0) {
```

Comment:
This condition determines if the size should be fetched from the context. I would add a comment explaining why a negative targetSize indicates this behavior.